### PR TITLE
Fixed error code difference occurring due to parallel worker enforced

### PR DESCRIPTION
**Pull request by Deepesh125**
### Description

When any error raised from parallel worker then Postgres by default does not share message_id (part of edata) to leader worker. Babelfish's error mapping framework relies on message_id to map to correct T-SQL error code. But since it is no more available to leader worker, T-SQL error code would not be found and default error code will be used. This will also affect the transactional behaviour of given error. 

In order to fix this issue, we need to store if given worker is spawned in the context of Babelfish or not to share error
message_id. Hence, this changes made changes in two data structure namely Port and FixedParallelState to store context
about Babelfish
 
### Issues Resolved
BABEL-4539

### Check List
Existing JDBC test cases should be sufficient to test these changes. 
Also, regression test suite provided by vanilla Postgres runs fine with these changes.

Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/libpq/pqmq.c [hunk 1]
```diff
@@ -310,6 +311,16 @@ pq_parse_errornotice(StringInfo msg, ErrorData *edata)
...
 				edata->funcname = pstrdup(value);
 				break;
+			case PG_DIAG_MESSAGE_ID:
+				if (sql_dialect == SQL_DIALECT_TSQL)
+				{
+					edata->message_id = pstrdup(value);
+				}
+				else
+				{
+					elog(ERROR, "Unexpected message_id found");
```
Comment (surendravishnoi) [inline src/backend/libpq/pqmq.c]: why are we giving error for APG here? also does this code applies to clients as well? they are not aware of SQL dialect

File: src/backend/utils/error/elog.c [hunk 2]
```diff
@@ -1703,6 +1704,8 @@ ThrowErrorData(ErrorData *edata)
 	if (edata->backtrace)
 		newedata->backtrace = pstrdup(edata->backtrace);
 	/* assume message_id is not available */
+	if (edata->message_id)
```
Comment (surendravishnoi) [inline src/backend/utils/error/elog.c]: why this change? message id is global at-least for given postgres worker

File: src/backend/utils/error/elog.c [hunk 3]
```diff
@@ -3250,6 +3253,15 @@ send_message_to_frontend(ErrorData *edata)
 			err_sendstring(&msgbuf, edata->funcname);
 		}
 
+		if (sql_dialect == SQL_DIALECT_TSQL)
+		{
+			if (edata->message_id)
```
Comment (surendravishnoi) [inline src/backend/utils/error/elog.c:3264]: how do we make sure that all clients will be able to handle additional message id in error messgage

File: src/backend/libpq/pqmq.c (hunk 1 above)
Comment (KushaalShroff) [reply src/backend/libpq/pqmq.c]: For PG anyway we were throwing `elog(ERROR, "unrecognized error field code: %d", (int) code);` before

Comment (Deepesh125) [reply src/backend/libpq/pqmq.c]: which clients are you referring to? Client here would be leader node (not psql that you might be referring here)

File: src/backend/utils/error/elog.c (hunk 2 above)
Comment (Deepesh125) [reply src/backend/utils/error/elog.c]: message_id is not global, its part of edata

File: src/backend/utils/error/elog.c (hunk 3 above)
Comment (Deepesh125) [reply src/backend/utils/error/elog.c:3264]: I think I will add another check (IsParallelWorker()) here to ensure that only parallel worker sends this message_id and client (leader node in this case) will be able to handle it through dialect.

File: src/backend/utils/error/elog.c (hunk 2 above)
Comment (surendravishnoi) [reply src/backend/utils/error/elog.c]: message ids are static strings, so there is one copy at process level. that is why we do not copy them, assigning pointer is sufficient. What is reasoning behind this change?

File: src/backend/utils/error/elog.c (hunk 3 above)
Comment (surendravishnoi) [reply src/backend/utils/error/elog.c:3264]: are we sure that parallel workers have consistent mapping with respect to dialect with leader node? what if error happens inside a procedure executed with PG dialect or vice versa and leader has different dialect?

File: src/backend/libpq/pqmq.c (hunk 1 above)
Comment (surendravishnoi) [reply src/backend/libpq/pqmq.c]: we should validate here as well that this is indeed in context of parallel worker (I am not sure if that is even possible). Also, code treats message id as static strings. If we are using str dup, we need to make sure that assumption is not broken when we simply assign pointers.

File: src/backend/utils/error/elog.c (hunk 2 above)
Comment (Deepesh125) [reply src/backend/utils/error/elog.c]: First copy is happening inside temporary context (which will be resetted at the end of handling parallel message) and second copy is happening inside error context which would be preserved at process level and will be used everywhere for error manipulation. So there is no harm in keeping this.

File: src/backend/utils/error/elog.c (hunk 3 above)
Comment (Deepesh125) [reply src/backend/utils/error/elog.c:3264]: As discussed offline, we can keep this message_id irrespective of dialect.

File: src/include/libpq/libpq-be.h [hunk 4]
```diff
@@ -250,6 +250,7 @@ typedef struct Port
 	SSL		   *ssl;
 	X509	   *peer;
 #endif
+	bool		is_tds_conn;
```
Comment (surendravishnoi) [inline src/include/libpq/libpq-be.h:253]: this information is available on TSQL side, we should add hook rather than adding a new member here.

File: src/backend/access/transam/parallel.c [hunk 5]
```diff
@@ -101,6 +101,9 @@ typedef struct FixedParallelState
 
 	/* Maximum XactLastRecEnd of any worker. */
 	XLogRecPtr	last_xlog_end;
+
+	/* Track if parallel worker is spawned in the context of Babelfish */
+	bool babelfish_context;
```
Comment (surendravishnoi) [inline src/backend/access/transam/parallel.c]: if this is not in the context of mutex, move it before mutex

Comment (Deepesh125) [reply src/backend/access/transam/parallel.c]: Thats good point, updated accordingly. Thanks

File: src/include/libpq/libpq-be.h (hunk 4 above)
Comment (Deepesh125) [reply src/include/libpq/libpq-be.h:253]: We rather keep it as it is for some other reasons.

**Review by surendravishnoi — APPROVED** [review]
